### PR TITLE
Enable Performance Insights and Autoscaling

### PR DESCRIPTION
**Pull request by rverma-nikiai**
## what
* Added support for enabling performance insight
* Added cluster autoscaling

## caveats
* not applicable for t2.small, t2.medium and t2.large.

### CODE REVIEW

**Comment by osterman** [comment]
Code needs formatting...
``
Terraform configuration needs linting. Run '/home/travis/build/cloudposse/terraform-aws-rds-cluster/build-harness/vendor/terraform fmt'
- main.tf 
- variables.tf`
```

File: main.tf [hunk 1]
```diff
@@ -76,6 +76,32 @@ resource "aws_rds_cluster_instance" "default" {
   engine_version          = "${var.engine_version}"
   monitoring_interval     = "${var.rds_monitoring_interval}"
   monitoring_role_arn     = "${var.rds_monitoring_role_arn}"
+  performance_insights_enabled        = "${var.performance_insights_enabled}"
+}
+  
+resource "aws_appautoscaling_target" "replicas" {
```
Comment (osterman) [inline main.tf]: Add option to disable autoscaling... perhaps `replicas_autoscaling_enabled`

Comment (osterman) [reply main.tf]: (set default to false so we don't cause changes for those who upgrade)

**Comment by osterman** [comment]
@rverma-nikiai looks like `variables.tf` still needs linting

File: main.tf [hunk 2]
```diff
@@ -64,18 +64,46 @@ resource "aws_rds_cluster" "default" {
...
+  monitoring_role_arn          = "${var.rds_monitoring_role_arn}"
+  performance_insights_enabled = "${var.performance_insights_enabled}"
+  availability_zone            = "${var.instance_availability_zones}"
+}
+
+resource "aws_appautoscaling_target" "replicas" {
+  service_namespace  = "rds"
+  scalable_dimension = "rds:cluster:ReadReplicaCount"
+  resource_id        = "cluster:${aws_rds_cluster.default.id}"
+  min_capacity       = 0
```
Comment (aknysh) [inline main.tf]: don't hardcode, make a variable for that

File: main.tf [hunk 3]
```diff
@@ -64,18 +64,46 @@ resource "aws_rds_cluster" "default" {
...
+  max_capacity       = 15
```
Comment (aknysh) [inline main.tf]: don't hardcode, make a variable for that

File: main.tf [hunk 4]
```diff
@@ -64,18 +64,46 @@ resource "aws_rds_cluster" "default" {
...
+  scalable_dimension = "${aws_appautoscaling_target.replicas.scalable_dimension}"
+  resource_id        = "${aws_appautoscaling_target.replicas.resource_id}"
+  policy_type        = "TargetTrackingScaling"
+
+  target_tracking_scaling_policy_configuration {
+    predefined_metric_specification {
+      predefined_metric_type = "RDSReaderAverageCPUUtilization"
+    }
+
+    target_value       = 75
```
Comment (aknysh) [inline main.tf]: don't hardcode, make variables for all these params

File: main.tf [hunk 5]
```diff
@@ -64,18 +64,46 @@ resource "aws_rds_cluster" "default" {
...
+resource "aws_appautoscaling_target" "replicas" {
+  service_namespace  = "rds"
+  scalable_dimension = "rds:cluster:ReadReplicaCount"
+  resource_id        = "cluster:${aws_rds_cluster.default.id}"
+  min_capacity       = 0
+  max_capacity       = 15
+}
+
+resource "aws_appautoscaling_policy" "replicas" {
+  name               = "cpu-auto-scaling"
```
Comment (aknysh) [inline main.tf]: the name could collide with other names. Use `label` ID

```hcl
name               = "${module.label.id}-cpu-auto-scaling"
```

File: variables.tf [hunk 6]
```diff
@@ -35,11 +35,16 @@ variable "subnets" {
   description = "List of VPC subnet IDs"
 }
 
-variable "availability_zones" {
+variable "cluster_availability_zones" {
   type        = "list"
   description = "List of Availability Zones that instances in the DB cluster can be created in"
 }
 
+variable "instance_availability_zones" {
```
Comment (aknysh) [inline variables.tf]: 1. instance AZ is single, not a list
https://www.terraform.io/docs/providers/aws/r/rds_cluster_instance.html#availability_zone

2. It's computed by AWS, not sure if we need to specify it

**Comment by osterman** [comment]
Needs another rebase too =)

**Comment by osterman** [comment]
@rverma-nikiai going through some of the open PRs and saw this wasn't updated =)

File: variables.tf (hunk 6 above)
Comment (rverma-nikiai) [reply variables.tf]: This was required if you don't want to distribute the autoscaling nodes within the azs. If you use default computed value, terraform will always recreate the rds instances with showing changes in azs.

**Comment by rverma-nikiai** [comment]
@osterman applied all the suggestions.

**Comment by aknysh** [comment]
thanks @rverma-nikiai 
all features from this PR were implemented in
https://github.com/cloudposse/terraform-aws-rds-cluster/releases/tag/0.9.0
https://github.com/cloudposse/terraform-aws-rds-cluster/releases/tag/0.10.0

We separated it into the two releases for easier tracking.
We'll close this PR for now.
thank you for your contribution.

**Pull request closed without merging**